### PR TITLE
Feature/channels and health checker

### PR DESCRIPTION
**Pull request by DavidSubiros**
### What

- pass channels from caller to producer and consumer creation functions.
- created interfaces and moq mocks for dependencies (sarama and sarama cluster*)
- add unit tests for consumer creation with channels
- add unit tests for producer creation with channels, and message, err, etc handling
- add Check() function for producer and consumer, which always return a successful check at the moment
- add unit tests for health checker
- modified example so that it does not die when Kafka service stops, and it doesn't panic on start if Kafka service is not running (expected behaviour in real systems that will use this library)

### How to review

- unit tests should pass
- try to run the example (under cmd/kafka-example) in the following scenarios:
  - Kafka service running -> example should work as expected
  - Kafka service stopped when example starts -> example should not panic
  - Kafka service running and example running. Then stop Kafka service -> example should log errors, but not die. If Kafka service starts again, I observed that Kafka will recover (desirable behaviour, but it was not a requirement originally)

### Who can review

anyone

* Note: Sarama cluster is deprecated, and we should use sarama for consumer-groups.

### CODE REVIEW

File: consumer-group.go [hunk 1]
```diff
@@ -57,10 +57,10 @@ func (cg *ConsumerGroup) StopListeningToConsumer(ctx context.Context) (err error
 	logData := log.Data{"topic": cg.topic, "group": cg.group}
 	select {
 	case <-cg.closed:
-		log.Info("StopListeningToConsumer got confirmation of closed kafka consumer listener", logData)
+		log.Event(nil, "StopListeningToConsumer got confirmation of closed kafka consumer listener", logData)
```
Comment (nshumoogum) [inline consumer-group.go]: Probably should pass a context value

File: consumer-group.go [hunk 2]
```diff
@@ -186,11 +217,11 @@ func newConsumer(brokers []string, topic string, group string, offset int64, syn
 		for looping := true; looping; {
 			select {
 			case <-cg.closer:
-				log.Info("Closing kafka consumer controller", logData)
+				log.Event(nil, "Closing kafka consumer controller", logData)
 				<-cg.closed
 				looping = false
 			case err := <-cg.consumer.Errors():
-				log.Error(err, nil)
+				log.Event(nil, "", log.Error(err))
```
Comment (nshumoogum) [inline consumer-group.go]: You should probably leave a generic message in error event, something like `kafka consumer error`

File: errors.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package kafka
+
+import "errors"
+
+// ErrShutdownTimedOut represents an error received due to the context
```
Comment (nshumoogum) [inline errors.go]: Comments for variables should be above each instantiation of variable unless the comment is generic to cover all variables in list (e.g. list of errors)

File: healthcheck.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+// getCheck creates a Check structure and populate it according the code, status and message
+func getCheck(ctx context.Context, code int, status, message string) *health.Check {
+
+	currentTime := time.Now().UTC()
+
+	check := &health.Check{
+		Name:        ServiceName,
+		Status:      status,
+		StatusCode:  code,
```
Comment (nshumoogum) [inline healthcheck.go]: Technically it shouldn't have a status code as this field is reserved for http requests

File: healthcheck_test.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+		}
+		chOut, chErr, chCloser, chClosed := createProducerChannels()
+		producer, err := kafka.NewProducerWithSaramaClient(testBrokers, testTopic, 123,
+			chOut, chErr, chCloser, chClosed, saramaCli)
+
+		So(err, ShouldBeNil)
+
+		Convey("Producer returns a successful Check structure", func() {
+			validateSuccessfulProducerCheck(&producer)
+			// So(len(producer.GetBucketReaderCalls()), ShouldEqual, 1)
```
Comment (nshumoogum) [inline healthcheck_test.go]: Should this be commented out?

File: healthcheck_test.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+		}
+		chUpstream, chCloser, chClosed, chErrors, chUpstreamDone := createConsumerChannels(true)
+		consumer, err := kafka.NewConsumerWithChannelsAndClusterClient(
+			testBrokers, testTopic, testGroup, kafka.OffsetNewest, true,
+			chUpstream, chCloser, chClosed, chErrors, chUpstreamDone, clusterCli)
+		So(err, ShouldBeNil)
+
+		Convey("Consumer returns a successful Check structure", func() {
+			validateSuccessfulConsumerGroupCheck(consumer)
+			// So(len(producer.GetBucketReaderCalls()), ShouldEqual, 1)
```
Comment (nshumoogum) [inline healthcheck_test.go]: And again?

File: producer.go [hunk 7]
```diff
@@ -46,47 +41,68 @@ func (producer *Producer) Close(ctx context.Context) (err error) {
 	case <-producer.closed:
 		close(producer.errors)
 		close(producer.output)
-		log.Info("Successfully closed kafka producer", nil)
+		log.Event(nil, "Successfully closed kafka producer")
```
Comment (nshumoogum) [inline producer.go]: Probably should create a context for logs

File: producer_test.go [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+// kafka topic that will be used for testing
+const testTopic = "testTopic"
+
+// timeout for test channels message propagation
+const TIMEOUT = 1 * time.Second
+
+// var testBrokers = []string{"1.1.1.1:1111", "2.2.2.2:2222"}
```
Comment (nshumoogum) [inline producer_test.go]: Comment left in for multiple brokers?

Do we need a test for multiple brokers?

File: consumer-group.go (hunk 1 above)
Comment (DavidSubiros) [reply consumer-group.go]: I guess this applies to all the log.Event(...) calls.

File: errors.go (hunk 3 above)
Comment (DavidSubiros) [reply errors.go]: Yes, I fixed this in the next PR. Maybe I can just remove this comment in this PR.

File: healthcheck.go (hunk 4 above)
Comment (DavidSubiros) [reply healthcheck.go]: Yes, fixed in next PR

File: healthcheck_test.go (hunk 5 above)
Comment (DavidSubiros) [reply healthcheck_test.go]: sorry, I forgot to remove it. This was from when I copied a similar function from the dp-s3 repo.

File: producer.go (hunk 7 above)
Comment (DavidSubiros) [reply producer.go]: Shall I provide the existing context if we have one, and create a logging one if it does not exist?

File: producer_test.go (hunk 8 above)
Comment (DavidSubiros) [reply producer_test.go]: I'll remove this comment from this PR. In the next PR, I created tests with 2 brokers, using a mock provided by Sarama. If I understand correctly, at the moment we are only using 1 broker, but the health check is prepared to handle a situation with multiple workers (try to contact all of them)

File: errors.go (hunk 3 above)
Comment (nshumoogum) [reply errors.go]: I would leave it as is if it is fixed in next pr

File: producer_test.go (hunk 8 above)
Comment (nshumoogum) [reply producer_test.go]: I think in the environments there maybe multiple brokers, not just one (default in apps)

**Pull request merged**